### DISCUSSION TITLE
Looking for support for PDF loading and storing feature

### DISCUSSION THREAD

**Discussion by stephanstapel** [stephanstapel/ZUGFeRD-csharp#470] [General]
Hi,

I have started porting the code from
https://github.com/Securibox/facturx/tree/main/Securibox.FacturX

into this library.

Therefore, I have created a dedicated project that I'm planning to put on nuget separately to the original package:
https://github.com/stephanstapel/ZUGFeRD-csharp/tree/master/ZUGFeRD.PDF

Three things need some help:
* Saving basically works, but resulting pdfs are not yet compliant. Can someone check the xmp metadata?
* If you have a better idea for a class name, I'd also be happy to discuss
* ~~Does anyone know how to express dependencies between nuget packages? I.e. the new component needs the ZUGFeRD-csharp component and also PdfSharp and I guess that needs to be put somewhere in the package description?!~~

**Comment by drocx** [comment]
How to express dependencies between nuget packages:

You define dependencies using `<PackageReference>` elements:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
    <PackageReference Include="Serilog" Version="[2.10.0,3.0.0)" />
  </ItemGroup>
</Project>
```


NuGet supports the following version range notations:

- `1.0.0`: Exact version.
- `[1.0.0]`: Exact version (same as above).
- `(1.0.0,)`: Greater than 1.0.0.
- `[1.0.0,2.0.0)`: Greater than or equal to 1.0.0 and less than 2.0.0.
- `[1.0.0,]`: Greater than or equal to 1.0.0.
- `*`: Accept any version.

https://learn.microsoft.com/en-us/nuget/concepts/dependency-resolution#dependency-resolution-with-packagereference

**Reply by stephanstapel** [reply]
Thanks! I have published an alpha version yesterday evening, For this first alpha, even Visual Studio dependencies worked:

![grafik](https://github.com/user-attachments/assets/55654389-9552-49c2-a0b2-58f04987a096)

If you find time for testing the code, I'd be more than happy to get your feedback!

**Comment by mhartmann-weissblau** [comment]
PDFSharp always overwrites the xmp metadata when saving the PDF file with their own values.

You can check this by setting a breakpoint to line
```
outputDocument.Save(memoryStream);
```
in method "_CreateFacturXStream" of class "InvoiceDescriptorPdfSaver" and check the content of the "/Metadata" entry.
After executing the method "Save" you will see, that value of the "/Metadata" entry was changed.

Overwriting the xmp metadata happens in the last line of method "PrepareForSave" of class "PdfDocument":
```
this.Catalog.Elements.SetReference("/Metadata", (PdfObject) new PdfMetadata(this));
```
There's an open pull request for PDFSharp for fixing this problem ([issue #44](https://github.com/empira/PDFsharp/pull/44)) since october 2023, but the developer(s) of PDFSharp want to have their own values in the xmp metadata.
Maybe you can switch to https://github.com/KDS/PDFsharp/tree/pdfsharp-extended (a fork of PDFSharp, where this issue is fixed).

**Reply by stephanstapel** [reply]
cool, thanks for letting me know, that is a good explanation why the pdf validation fails. I dropped a question on that issue, let's see.


Eight commits behind :(

![grafik](https://github.com/user-attachments/assets/53fc8256-47eb-4295-b1d2-72c0f935aff9)

**Reply by mhartmann-weissblau** [reply]
I agree: 8 commits can fix a lot of bugs. 

However, it seems that the developers of PDFSharp are not willing to change this behavior regarding overwriting of xmp metadata and therefore it may be impossible to create a valid ZUGFeRD PDF file with PDFSharp. 
From my point of view PDFSharp needs some kind of merging existing xmp metadata with own xmp metadata, but that's not so easy.

**Reply by stephanstapel** [reply]
The fork works like a charm. There seems to be just one last issue:

![grafik](https://github.com/user-attachments/assets/5bc34260-381f-4af6-a8f8-f98dd65d72fc)

any idea?

**Reply by m4Viper** [reply]
"Signatur: unknown" ist not the reason for "nicht valide"

I am creating the xml (ZUGFeRDVersion.Version23, Profile.Extended) with ZUGFeRD-csharp and then I save it together with the pdf (created with iText, pdf/A-3A)

All my pdf´s are "valide", but this "signatur: unknown" message pops up. -> So this message is not the reason for your not being valid problem.
![pdf](https://github.com/user-attachments/assets/50dfcac0-6856-4d90-9880-37ba01999b45)

**Reply by stephanstapel** [reply]
@m4Viper : ok,thanks. Is your opinion that I can ignore this message?

**Reply by m4Viper** [reply]
Ich habe heute eine ZUGFeRD-PDF Rechnung von einem Lieferanten erhalten und gleich mal validiert.
Die XML ist korrekt aber die pdf-Verschmelzung scheint auch fehlerhaft zu sein. (DATEV akzeptiert sie aber und liest sie aus)
Es kommt der gleiche Fehler wie bei dir:

_Das ZUGFeRD-PDF ist nicht valide.
Signatur: unknown

Details_

Wenn ich mir die PDF genauer ansehe sehe ich, dass z.B.: die nötigen pdf-Eigenschaften wie "isTaged", "Language", "Title" nicht gesetzt sind.
Die XMP metadata sind auch nicht vollständig.

Eventuell hat dein erstelltes PDF ein ähnliches Problem.

**Reply by tdreiner** [reply]
Hallo zusammen, 

gibt es zu dem Signatur: unknown mittlerweile irgendwelche Infos, wie ich das wegbekomme? Das ist die einzige Meldung, die ich noch bekomme. Ich signiere unsere PDF mit einem gültigen und für Adobe zulässigen Zertifikat.

**Reply by IsThatUniqueEnough** [reply]
> gibt es zu dem Signatur: unknown mittlerweile irgendwelche Infos, wie ich das wegbekomme?

Das ist keine Validierungsmeldung, sondern eine reichlich unglücklich benamste Information. Mustang versucht anhand von neun Magic Strings festzustellen, mit welcher Software die E-Rechnung erzeugt wurde. Findet es keine dieser "Signaturen", wird "unknown" ausgegeben. Mit kryptografischen Signaturen hat das nichts zu tun.

**Reply by stephanstapel** [reply]
https://github.com/ZUGFeRD/mustangproject/blob/17bc5c483f83b83e8ecf18f2109bf4e3fdc4881c/validator/src/main/java/org/mustangproject/validator/PDFValidator.java#L262
